utilities/ggd/main.go

```diff
@@ -0,0 +1,318 @@
+ggd branchA..branchB
+
+# (4) Run tests for all the golang packages affected by between master, and head
+go test $(ggd)
+
+# (5) List all the golang packages affected by between master and head, and build tags 'integration'
+ggd -t integration
+
+# (6) The same as (5), but include debug output (sent to STDERR)
+ggd -t integration -d
+
+# (7) The same as (6), but include debug output (sent to STDERR), and
+# save the generated DAG in changes.png for visualization
+ggd -t integration -d -o change.dot
+dot -Tpng change.png change.dot
+`
+
+var (
+	basePkg           = getopt.StringLong("base-package", 'p', "", "repository package (can leave blank typically)")
+	includePrechanges = getopt.BoolLong("include-prechange", 'i', "include prechanges for golang import dag calculation")
+	debugMode         = getopt.BoolLong("debug", 'd', "debug mode - useful for understanding output")
+	buildTags         = getopt.ListLong("build-tags", 't', "comma separated golang build tags")
+	dotOutputFile     = getopt.StringLong("output-file", 'o', "",
+		"optionally, visualize changes in a dot file (only in debug mode)")
+)
+
+// Run executes the ggd command.
+func main() {
+	getopt.Parse()
+	if err := validateArgs(); err != nil {
+		printUsage()
+		die("illegal arguments: %v", err)
+	}
+	sha1Fn := lib.CurrentSHA1
+
+	baseRef, headRef, err := rangeExtractor{sha1Fn}.extractRanges(getopt.Args())
+	dieIfErr(err, "unable to extract ranges: %v", err)
+	commitRange := fmt.Sprintf("%s..%s", baseRef, headRef)
+	debug("retrieving files changed from git in commit-range: %v", commitRange)
+
+	gitChanges, err := lib.ChangedFiles(commitRange, *basePkg)
+	dieIfErr(err, "unable to find changed files from git: %v", err)
+
+	debug("files changed from git: %v", gitChanges)
+	debug("calculating changed packages")
+	packageChanges := lib.ChangedPackages(gitChanges)
+	debug("changed packages: %v", packageChanges)
+
+	baseFullChangeSet := make(lib.ImportSet)
+	if !*includePrechanges {
+		debug("skipping pre-change affected packages (includePrechanges=false)")
+	} else {
+		dirty, err := lib.CwgIsDirty()
+		dieIfErr(err, "unable to checkout base-branch for prechange impact: %v", err)
+		dieIf(dirty, "cwg is dirty, unable to checkout base-branch for prechange impact")
+
+		// checkout base branch, compute affected packages
+		debug("computing pre-change affected packages", baseRef)
+		debug("checking out base-branch: %v", baseRef)
+
+		checkoutResult, err := lib.Checkout(baseRef)
+		dieIfErr(err, "unable to checkout master: %v", err)
+
+		debug("checked out base-branch: %v", baseRef)
+		debug("git-output: %v", checkoutResult)
+		debug("calculating affected packages on base-branch due to changes")
+
+		_, _, baseFullChangeSet, err = dagHelper(packageChanges, *basePkg)
+		dieIfErr(err, "unable to compute affected packages: %v", err)
+		debug("affected packages (including transitive changes): %v", baseFullChangeSet)
+	}
+
+	currentSha1, err := sha1Fn()
```

Comment:
Should we try to checkout the current SHA again before we return?

also, super nit: `currentSha1` -> `currentSHA1`